File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.py
"""
"""
import sys

input = lambda: sys.stdin.readline().strip()

N = int(input())

def hanoi(cnt, source, target, left):
    if cnt == 0:
        return 0
    result = 0
    result += hanoi(cnt - 1, source, left, target)
    lst.append(str(source) + " " + str(target))
    result += hanoi(cnt - 1, left, target, source)
    return result + 1

lst = []
answer = hanoi(N, 1, 3, 2)
print(answer)
for a in lst:
    print(a)

File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        int N = Integer.parseInt(br.readLine());

        int answer = hanoi(N, 1, 3, 2);

        System.out.println(answer);
        System.out.println(sb.toString());

    } // end of main

    private static int hanoi(int cnt, int from, int to, int left) {
        if (cnt == 0) return 0;
        int result = 0;
        result += hanoi(cnt - 1, from, left, to);
        sb.append(from + " " + to + "\n");
        result += hanoi(cnt - 1, left, to, from);
        return result + 1;
    }

} // end of class